File: interview-preparation/week-1/ruby/09_pangram.rb
#!/bin/ruby

require 'json'
require 'stringio'
require 'set'

#
# Complete the 'pangrams' function below.
#
# The function is expected to return a STRING.
# The function accepts STRING s as parameter.
#

def pangrams(s)
  # Write your code here
  # Convert the string to lowercase for case-insensitive comparison
  s = s.downcase

  # Create a set to keep track of the presence of each letter
  letter_set = Set.new

  # Iterate over each character in the string
  s.each_char do |ch|
    # Check if the character is a letter
    if ch.match?(/[a-z]/)
      # Add the letter to the set
      letter_set.add(ch)
    end
  end

  # Check if all letters from 'a' to 'z' are present
  ('a'..'z').each do |ch|
    return 'not pangram' unless letter_set.include?(ch)
  end

  'pangram'
end

fptr = File.open(ENV['OUTPUT_PATH'], 'w')

s = gets.chomp

result = pangrams s

fptr.write result
fptr.write "\n"

fptr.close()
#!/bin/ruby

require 'json'
require 'stringio'
require 'set'

#
# Complete the 'pangrams' function below.
#
# The function is expected to return a STRING.
# The function accepts STRING s as parameter.
#

def pangrams(s)
  # Convert the string to lowercase for case-insensitive comparison
  s = s.downcase

  # Create a set to keep track of the presence of each letter
  letter_set = Set.new

  # Iterate over each character in the string
  s.each_char do |ch|
    # Check if the character is a letter
    if ch.match?(/[a-z]/)
      # Add the letter to the set
      letter_set.add(ch)
    end
  end

  # Check if all letters from 'a' to 'z' are present
  ('a'..'z').each do |ch|
    return 'not pangram' unless letter_set.include?(ch)
  end

  'pangram'
end

fptr = File.open(ENV['OUTPUT_PATH'], 'w')

s = gets.chomp

result = pangrams s

fptr.write result
fptr.write "\n"

fptr.close()

File: interview-preparation/week-1/go/09_pangram.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"
)

/*
 * Complete the 'pangrams' function below.
 *
 * The function is expected to return a STRING.
 * The function accepts STRING s as parameter.
 */

func pangrams(s string) string {
	// Convert the string to lowercase for case-insensitive comparison
	s = strings.ToLower(s)

	// Create a map to keep track of the presence of each letter
	letterMap := make(map[rune]bool)

	// Iterate over each character in the string
	for _, ch := range s {
		// Check if the character is a letter
		if ch >= 'a' && ch <= 'z' {
			// Mark the letter as present in the map
			letterMap[ch] = true
		}
	}

	// Check if all letters from 'a' to 'z' are present
	for ch := 'a'; ch <= 'z'; ch++ {
		if !letterMap[ch] {
			return "not pangram"
		}
	}

	return "pangram"
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	stdout, err := os.Create(os.Getenv("OUTPUT_PATH"))
	checkError(err)

	defer stdout.Close()

	writer := bufio.NewWriterSize(stdout, 16*1024*1024)

	s := readLine(reader)

	result := pangrams(s)

	fmt.Fprintf(writer, "%s\n", result)

	writer.Flush()
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}
